README.md
## phantomjs-remote

Run [PhantomJS](http://phantomjs.org/) scripts on a remote server (e.g. if the phantomjs
binary is not available on the local system, like a PaaS virtual machine with an
unsupported architecture for [phantomjs node wrapper](https://github.com/Obvious/phantomjs).)

This is my first node.js script, go easy on me :)

## Usage

```js
var phantomjs_remote = require('phantomjs-remote');
phantomjs_remote.send(phantomJsScript, callback, [options], [host], [port]).
```

*options* is an optional JSON dictionary with any required options, for now only *load-images* is supported, e.g. { 'load-images': 'no' }.

The last two options may be ommitted if PHANTOMJS_REMOTE environment variable is set
with "host:port".

Short example:
```js
phantomjs_remote('console.log(1); phantom.exit();', function(error, result) {
	if (!error)
		console.log(result);
}, { 'load-images': 'no'} );
```

## Server script / SECURITY

You are advised against running this script on it's own and should instead:

1. Run it in a chroot jail
2. Firewall the port to only receive connections from trusted hosts

Anyways, run like this:

```bash
node phantom-server.js <listenPort>
```

or
```bash
while `true`; do node phantom-server.js <listenPort> ; done 
```



History.md
# phantomjs-remote changelog

## 0.0.9

* Fix duplicate callback call on error (fixes #1)

phantom-repl.js
/*
 * PhantomJS proxy v0.1, (c) 2013 Gadi Cohen, released under the GPLv3.
 *
 * Listen on a port, receive entire scripts at a time, run them on phantomjs and return
 * the output to the sending socket.  I prefer this to forking a new instance of
 * phantomjs for each request, as that wastes time and system resources.
 *
 * This is my first node script :)  Go easy on me.  I got help from:
 * http://www.davidmclifton.com/2011/07/22/simple-telnet-server-in-node-js/
 *
 *
 * TODO:
 *
 * [] Ability to serve up static files created by phantomjs.  I don't need this for my
 *    purposes but I guess it would be useful.
 */ 

var net = require('net');
 
var spawn = require('child_process').spawn;
var phantomjs = spawn('phantomjs');

var sockets = [];
var queue = [];
var buffer = '';
var buffers = {};

phantomjs.stdout.on('data', function(data) {
	console.log('ph: ' + data.toString());
	if (data == "phantomjs> ") {
		if (buffer == '')
			return;

		var matches = /!!({.*})!!/.exec(buffer);
		if (matches) {
			var obj = JSON.parse(matches[1]);
			console.log(obj);
			if (obj.exit)
				sockets[obj.socket].end(obj.data);
			else
				sockets[obj.socket].write(obj.data);
		} else {
			console.log("Didn't know what to do with the following buffer:");
			console.log(buffer);
		}

		var socket = queue.shift();
		if (sockets.indexOf(socket) != -1)
			socket.end(buffer);
		buffer = '';
	} else if (data.toString().substr(0,11) == "phantomjs> ") {
		buffer += data.toString().substr(0, data.length-11);
	} else {
		buffer += data;
	}
});

phantomjs.stderr.on('data', function (data) {
    console.log('PhantomJS error: ' + data);
});

phantomjs.on('close', function (code) {
    console.log('PhantomJS exited with code ' + code);
});


/*
 * Take first line of input (there should only be one, and it should be shorter
 * than node's buffer... this is fine for spiderable, untested with anything else.
 */
function receiveData(socket, data) {
	var str = data.toString().replace(/\r|\n/g, '');
	var i = sockets.indexOf(socket);

	//console.log('[socket:'+i+'] "' + str + '"');
	if (str.match(/_EOF_$/)) {
		queue.push(socket);
		buffers[i] = '(function(){'
			+ 'function remote_return(data, exit) {'
			+ '  console.log("!!"+JSON.stringify({ socket: ' + i + ', data: data, exit: exit })+"!!");'
			+ '} ' + buffers[i] + str.replace(/_EOF_$/, '') + '}).call(this);\n';
		console.log('sending: ' + buffers[i]);
		phantomjs.stdin.write(buffers[i]);
	} else {
		buffers[i] += str;
	}
}

function closeSocket(socket) {
	var i = sockets.indexOf(socket);
	if (i != -1) {
		sockets.splice(i, 1);
	}
}

function newSocket(socket) {
	console.log('[' + new Date().toLocaleString() + '] New connection');
	sockets.push(socket);
	buffers[sockets.indexOf(socket)] = '';
	socket.on('data', function(data) {
		receiveData(socket, data);
	});	
	socket.on('end', function() {
		closeSocket(socket);
	});
}
 
// Create a new server and provide a callback for when a connection occurs
var server = net.createServer(newSocket);
 
// Listen on port 8888
server.listen(8000);


phantom-client.js
var net = require('net');

var socket = new net.Socket();
var callback;
var buffer;

exports.send = function(script, _callback, options, host, port) {
    if (!(host && port)) {
        if (!process.env.PHANTOMJS_REMOTE) {
            console.log('phantomjs_remote: Set PHANTOMJS_REMOTE environment variable or call with ,host,port');
            return;
        } else {
            host = process.env.PHANTOMJS_REMOTE.split(':');
            port = host[1]; host = host[0];
        }
    }
    buffer = '';
    callback = _callback;
    socket.connect(port, host, function() {
//        console.log('Connected to ' + host + ':' + port);
        script = (options ? 'OPTIONS:' + JSON.stringify(options) + ' ' : '')
               + script + '_EOF_';
        socket.write(script);
    }); 
};

socket.on('data', function(data) {
    buffer += data.toString();
});

socket.on('error', function(exception){
    if (callback) {
        callback(exception);
        callback = null;  // Avoid repeat call on close
    }
    // console.log(exception);
});


socket.on('drain', function() {
    // console.log("drain!");
});

socket.on('timeout', function() {
    if (callback) {
        callback('timeout');
        callback = null;  // Avoid repeat call on close
    }
    // console.log("timeout!");
});

// Add a 'close' event handler for the client socket
socket.on('close', function() {
    if (callback) {
        if (buffer.match(/^ERROR: /))
            callback(buffer);
        else
            callback(false, buffer);
    }
    // console.log('Connection closed, buffer sent to callback');
});


phantom-server.js
/*
 * PhantomJS proxy v0.1, (c) 2013 Gadi Cohen, released under the GPLv3.
 *
 * Accepts scripts to run in PhantomJS over a network connection 
 *
 * This is my first node script :)  Go easy on me.  I got help from:
 * http://www.davidmclifton.com/2011/07/22/simple-telnet-server-in-node-js/
 *
 * The original plan was to keep a single phantomjs running in REPL mode rather than
 * spawning a new process for each script, but it seems REPL mode is broken, i.e.
 * https://github.com/ariya/phantomjs/issues/11180
 *
 * Once the REPL is rewritten/fixed, I intend to continue work on the old code (see
 * phantomjs-repl.js), as I think this will be faster and less resource intensive.
 *
 * TODO:
 *
 * [] Ability to serve up static files created by phantomjs.  I don't need this for my
 *    purposes but I guess it would be useful.
 */ 


var net = require('net');
var child_process = require('child_process');

if (process.argv.length < 3) {
	console.log('Usage: ./phantom-server <port>');
	return;
}

var port = process.argv[2];
var sockets = [];
var buffer = '';
var buffers = [];

// how long to let phantomjs run before we kill it
var REQUEST_TIMEOUT = 15*1000;

// only allow safe user input
function process_options(options) {
	var out = {};
	if (options['load-images'])
		out['load-images'] = (options['load-images'] == 'yes') ? 'yes' : 'no';
	return out;
}

function receiveData(socket, data) {
	var str = data.toString().replace(/\r|\n/g, '');
	var i = sockets.indexOf(socket);
	var options = {};

	if (str.match(/_EOF_$/)) {
		var script = buffers[i] + str.replace(/_EOF_$/, '');
		script = script.replace(/^OPTIONS:(\{.*?\}) /, function(match, _options) {
			options = process_options(JSON.parse(_options));
			return '';
		});

		// Security note: The script has had all newlines stripped.  Our script
		// is piped through bash, so this also ensures that no one can send a
		// newline followed by _EOF_, which would allow them to run custom shell
		// commands (bash injection).

	    // Run phantomjs (this note taken from Meteor spiderable package)
	    //
	    // Use '/dev/stdin' to avoid writing to a temporary file. We can't
	    // just omit the file, as PhantomJS takes that to mean 'use a
	    // REPL' and exits as soon as stdin closes.
	    //
	    // However, Node 0.8 broke the ability to open /dev/stdin in the
	    // subprocess, so we can't just write our string to the process's stdin
	    // directly; see https://gist.github.com/3751746 for the gory details. We
	    // work around this with a bash heredoc. (We previous used a "cat |"
	    // instead, but that meant we couldn't use exec and had to manage several
	    // processes.)

		var execLine = 'exec phantomjs';
		if (options['load-images'])
			execLine += " --load-images=" + options['load-images'];
		execLine += ' /dev/stdin << _EOF_\n' + script + '\n_EOF_\n';

	    child_process.execFile('/bin/bash', ['-c', (execLine)],
	      {timeout: REQUEST_TIMEOUT},
	      function (error, stdout, stderr) {
	        if (!error) {
	        	if (sockets.indexOf(socket) != -1)
	        		socket.end(stdout);
	        } else {
				if (error && error.code === 127) {
					socket.end("ERROR: phantomjs not installed. Download and install from http://phantomjs.org/");
					console.log("phantomjs not installed. Download and install from http://phantomjs.org/");
				} else {
					socket.end("ERROR: phantomjs failed: " +  error + "\nstderr:" + stderr);
					console.log("phantomjs failed:", error, "\nstderr:", stderr);
				}
	        }
	      }
	    );
	} else {
		buffers[i] += str;
	}
}

function closeSocket(socket) {
	var i = sockets.indexOf(socket);
	if (i != -1) {
		sockets.splice(i, 1);
	}
}

function newSocket(socket) {
	console.log('[' + new Date().toLocaleString() + '] New connection from '
		+ socket.remoteAddress);
	sockets.push(socket);
	buffers[sockets.indexOf(socket)] = '';
	socket.on('data', function(data) {
		receiveData(socket, data);
	});	
	socket.on('end', function() {
		closeSocket(socket);
	});
}
 
var server = net.createServer(newSocket);
server.listen(port);
